src/urllib3/connection.py

```diff
@@ -251,40 +251,6 @@ def __init__(
         # HTTPS requests to go out as HTTP. (See Issue #356)
         self._protocol = "https"
 
-    def connect(self):
```

Comment:
We probably shouldn't remove this, what happens if a user has an unverified HTTPS connection, is this class used?

Reply:
No, it's not used, unverified connections go through VerifiedHTTLS connection

Reply:
Oh, huh. I wonder what the impact of nuking the class and rebinding the name would do to downstream. :man_shrugging:

Reply:
@sethmlarson I opened https://github.com/urllib3/urllib3/issues/1730 to unblock this pull request and talk about our options here, including the upstream impact.

Reply:
Could we try running vcrpy's test suite on this branch to ensure removing this function doesn't break anything?

Reply:
I can confirm that the vcrpy test suite continue to work with this change, both for requests and urllib3

Reply:
> No, it's not used, unverified connections go through VerifiedHTTPSConnection.

You mean, it was not covered by the existing test base and you thought it was a good case for its removal ?

Reply:
@zogzog Sorry for causing an issue. I've edited your above comment as I read it as sarcastic which isn't constructive. Hope you agree with my edit.

Reply:
> I don't think anybody is using this class: connect() referenced variables that don't exist, so could not work. 

Weird statement. My code definitely go through this path.
Indeed I had to add

    cert_reqs = 'CERT_NONE'
    ca_certs = None
    ca_cert_dir = None

6 months ago because someone removed them or something, and urllib3 doesn't have coverage for this indeed. This not the correct way to kill a feature though.

I think it came with https://github.com/urllib3/urllib3/pull/1608/commits/6c6196717eaa29ffb2269037762f958cb4f6ed79 

> An additional wrinkle here is the bottom of the file:
> 
> ```
> if ssl:
>     # Make a copy for testing.
>     UnverifiedHTTPSConnection = HTTPSConnection
>     HTTPSConnection = VerifiedHTTPSConnection
> ```
> 
> So one option is to merge VerifiedHTTPSConnection into HTTPSConnection, and add `UnverifiedHTTPSConnection = HTTPSConnection` at the end of the file.

People should really understand that, even though it is generally considered a bad thing and should be discouraged, there is a need for:

* unverified https connections,
* that don't spit warnings on the console.

I have to maintain something very complicated like this

```python

# HTTPS verfication/un-warning

# HTTPS warning disabled for mercure requests
# We don't want to verify the Mercure certificate because
# it is too complicated to maintain.

# We cannot disable this warning using standard catch_warning
# tricks because they are not thread-safe,
# hence we go to greater lengths to *not emit* the warning.

# We monkey-patch HTTPSConnectionPool._validate_conn to make it
# recognize a new special case, and provide a Session object
# which triggers the special case.


def _validate_conn(self, conn):
    """
    Called right before a request is made, after the socket is created.
    """
    # Force connect early to allow us to validate the connection.
    if not getattr(conn, 'sock', None):  # AppEngine might not have  `.sock`
        conn.connect()

    # PATCH
    if isinstance(conn, SilentUnverifiedHTTPSConnection):
        return
    # /PATCH

    if not conn.is_verified:
        warn((
            'Unverified HTTPS request is being made. '
            'Adding certificate verification is strongly advised. See: '
            'https://urllib3.readthedocs.io/en/latest/advanced-usage.html'
            '#ssl-warnings'),
            requests.packages.urllib3.exceptions.InsecureRequestWarning)


requests.packages.urllib3.connectionpool.HTTPSConnectionPool._validate_conn = _validate_conn


class SilentUnverifiedHTTPSConnection(
        requests.packages.urllib3.connection.UnverifiedHTTPSConnection):
    # the only significant information here is a connection of a
    # specific type, sufficient for an isinstance check

    cert_reqs = 'CERT_NONE'
    ca_certs = None
    ca_cert_dir = None


class NoHTTPSVerifyAdapter(requests.sessions.HTTPAdapter):

    def cert_verify(self, cn, *_args):
        # we override this (class) attribute on the fly
        # (cn is actually a connection pool)
        cn.ConnectionCls = SilentUnverifiedHTTPSConnection


class NoHTTPSVerifySession(requests.Session):

    def __init__(self):
        super(NoHTTPSVerifySession, self).__init__()
        # we redo the adapters mapping with our custom
        # silence-inducing adapters
        self.adapters = OrderedDict()
        self.mount('https://', NoHTTPSVerifyAdapter())
        self.mount('http://', NoHTTPSVerifyAdapter())

```

I would LOVE getting rid of this.

Reply:
Aurélien, je comprends ta colère, je sais ce que c'est de tomber sur un commit quelque part sur GitHub qui casse ton code. Je te présente d'ailleurs mes excuses. Maintenant que ça c'est dit, on apprécierait des interventions plus courtoises !

> Indeed I had to add [...] 6 months ago because someone removed them or something

We would have appreciated a heads-up at this point! This would have avoided the whole issue.

I'd be happy to accept a pull request that restores a working and covered connect().

> People should really understand that, even though it is generally considered a bad thing and should be discouraged, there is a need for unverified https connections that don't spit warnings on the console.

Can you use our fingerprinting feature if the certificate does not change too often?

Is silencing the warning an option?

By the way, it looks like the TLS certificate of https://pythonian.fr/ is currently expired.

Reply:
Sorry for venting my frustration, I didn't wait long enough to let it wind down before writing here. Now a constructive interaction there is all we need :)

6 months ago I just didn't have the time to do a root cause analysis of our issues. Priorities, etc. That's life. 

The whole point about unverified ssl (in our case, I guess it is a real issue in the corporate world in many places) is that we can't afford to break our tools if the IT department forgets to update a certificate or even uses a self-signed thing on some pre-production instance.

I have nothing against unverified https connection warning _by default_ but having a "don't warn, I know what I am doing" flag would be appreciated.

The tool has a command line for casual queries and we don't want the end-users to be pestered about the "unsafe" connection. We are basically powerless against it.

So yes, silencing the warning should be an option imho.

Reply:
I mean, why don't you just silence the warning? Isn't this working for you?

```python3
import urllib3

urllib3.disable_warnings(urllib3.exceptions.SecurityWarning)
http = urllib3.PoolManager(cert_reqs="CERT_NONE")
r = http.request("GET", "https://pythonian.fr/webfiles/css.css")
print(r.status)
```

Reply:
Also, for what it's worth, here are the thoughts of Seth on verifying HTTPS: https://sethmlarson.dev/blog/2019-11-26/designing-for-real-world-https. I happen to fully agree.

Reply:
There's also this method (using certificate pinning as mentioned above) that doesn't require you to disable security warnings. :)

```python
http = urllib3.PoolManager(
    cert_reqs="CERT_NONE",
    assert_fingerprint="e677022f45ad8ecfc7a66dbb8d57e9c7b01a52c9f880ecb104af6ee67ee9f18a"
)
resp = http.request("GET", "https://pythonian.fr/webfiles/css.css")
```

Just remember this PoolManager will only work for the one website. Might be worth even using a ConnectionPool instead.

Reply:
> I mean, why don't you just silence the warning? Isn't this working for you?
> 
> ```python
> import urllib3
> 
> urllib3.disable_warnings(urllib3.exceptions.SecurityWarning)
> http = urllib3.PoolManager(cert_reqs="CERT_NONE")
> r = http.request("GET", "https://pythonian.fr/webfiles/css.css")
> print(r.status)
> ```

Looks like it globally disables warnings, right ? We don't want that !

Reply:
> Also, for what it's worth, here are the thoughts of Seth on verifying HTTPS: https://sethmlarson.dev/blog/2019-11-26/designing-for-real-world-https. I happen to fully agree.

I know all of this. Yet this does not seem to address all the real world issues. We are walking a fine line there between the need to operate securely by default and the need to make things work in the face of a failing security setup. Operational constraints make all the propositions there, though well-meant, not workable in practice.

Reply:
> There's also this method (using certificate pinning as mentioned above) that doesn't require you to disable security warnings. :)
> 
> ```python
> http = urllib3.PoolManager(
>     cert_reqs="CERT_NONE",
>     assert_fingerprint="e677022f45ad8ecfc7a66dbb8d57e9c7b01a52c9f880ecb104af6ee67ee9f18a"
> )
> resp = http.request("GET", "https://pythonian.fr/webfiles/css.css")
> ```
> 
> Just remember this PoolManager will only work for the one website. Might be worth even using a ConnectionPool instead.

This fingerprint, which is a piece of data related to the _current_ possibly broken certificate would cease to operate when the certificate is renewed right ? Also I would have to either:
* hardcode it and then the software would need to be re-released if the upstream certificate changed
* made a config item, whose actual value would have to be transmitted to end users in the case of change

Am I missing something ? Either way this is not practical enough.